openshift-client/src/main/java/io/fabric8/openshift/client/internal/OpenShiftOAuthInterceptor.java

```diff
@@ -16,124 +16,130 @@
-      } else {
-        return response;
-      }
+      setAuthHeader(builder, token);
+      request = builder.build();
+      return chain.proceed(request); //repeat request with new token
+    } else {
+      return response;
     }
+  }
 
-    private void setAuthHeader(Request.Builder builder, String token) {
-        if (token != null) {
-            builder.header(AUTHORIZATION, String.format("Bearer %s", token));
-        }
+  private void setAuthHeader(Request.Builder builder, String token) {
+    if (token != null) {
+      builder.header(AUTHORIZATION, String.format("Bearer %s", token));
     }
+  }
 
-    private String acquireNewToken() throws IOException {
-      synchronized (client) {
-        // current token (if exists) is broken, don't resend
-        oauthToken.set(null);
-        String token = authorize();
-        if (token != null) {
-          oauthToken.set(token);
-        }
+  private  String authorize() {
+    try {
+      OkHttpClient.Builder builder = client.newBuilder();
+      builder.interceptors().remove(this);
+      OkHttpClient clone = builder.build();
+
+      URL url = new URL(URLUtils.join(config.getMasterUrl(), AUTHORIZATION_SERVER_PATH));
+      Response response = clone.newCall(new Request.Builder().get().url(url).build()).execute();
+
+      if (!response.isSuccessful() || response.body() == null) {
+        throw new KubernetesClientException("Unexpected response (" + response.code() + " " + response.message() + ")");
       }
-      return oauthToken.get();
-    }
 
-    private  String authorize() throws IOException {
-        OkHttpClient.Builder builder = client.newBuilder();
-        builder.interceptors().remove(this);
-        OkHttpClient clone = builder.build();
-
-        String credential = Credentials.basic(config.getUsername(), new String(config.getPassword()));
-        URL url = new URL(URLUtils.join(config.getMasterUrl(), AUTHORIZE_PATH));
-        Response response = clone.newCall(new Request.Builder().get().url(url).header(AUTHORIZATION, credential).build()).execute();
-
-        response.body().close();
-        response = response.priorResponse() != null ? response.priorResponse() : response;
-        response = response.networkResponse() != null ? response.networkResponse() : response;
-        String token = response.header(LOCATION);
-        if (token == null || token.isEmpty()) {
-          throw new IOException("Unexpected response(" + response.code() + " " + response.message() + "), to the authorization request. Missing header:[" + LOCATION + "]!");
-        }
-        token = token.substring(token.indexOf(BEFORE_TOKEN) + BEFORE_TOKEN.length());
-        token = token.substring(0, token.indexOf(AFTER_TOKEN));
-        return token;
+      String body = response.body().string();
+      JsonNode jsonResponse = Serialization.jsonMapper().readTree(body);
+      String authorizationServer = jsonResponse.get("authorization_endpoint").asText();
+      response.body().close();
```

Comment:
Doesn't using `response.body().string()` closes body itself?

Reply:
Not sure, but I know that `response.close()` does